server/src-lib/Hasura/RQL/DDL/RemoteSchema.hs

```diff
@@ -94,10 +107,38 @@ removeRemoteSchemaP2
   => RemoteSchemaName
   -> m EncJSON
 removeRemoteSchemaP2 rsn = do
+  dropRemoteSchemaDirectDeps rsn
   delRemoteSchemaFromCache rsn
   liftTx $ removeRemoteSchemaFromCatalog rsn
   return successMsg
 
+dropRemoteSchemaDirectDeps
+  :: ( CacheRWM m
+     , MonadTx m
+     )
+  => RemoteSchemaName
+  -> m ()
+dropRemoteSchemaDirectDeps rsn = do
+  sc <- askSchemaCache
+  let depObjs = getDependentObjs sc objId
+      permDeps = catMaybes $ map getPermObj depObjs
+  mapM_ dropRemoteSchemaPermDep permDeps
+  where
+    objId = SORemoteSchema rsn
+    getPermObj = \case
+      obj@(SORemoteSchemaObj _(RSOPerm _)) -> Just obj
+      _ -> Nothing
```

Comment:
I'd prefer to throw `500 Internal error` with `unexpected dependent object`

Reply:
Oh! It's there in `dropRemoteSchemaPermDep`. It's better to `mapM_ dropRemoteSchemaPermDep` on `depObjs` directly and void `getPermObj`.